### ISSUE TITLE
Large Variability Across Seeded Experiments with CTC Loss

### ISSUE THREAD

**Issue by josecruzado21** [espnet/espnet#6433]
Hi all,

I am currently running experiments using the MLSUPERB setup in ESPnet (based on the 202511 version of the codebase), and I would like to ask about the level of variability I am observing across runs, even when I control for randomness using the same seed. The code for my implementation can be found here: https://github.com/josecruzado21/espnet_ctc_dro/tree/ml-superb-experiments

My setup is very close to the default pipeline, with only minor modifications to restrict training to a smaller subset of languages and datasets. The configuration is as follows:

- Data: Subset of MLSUPERB2 
- Languages and datasets: English (VoxForge), German (VoxForge), Hebrew (FLEURS), Japanese (FLEURS), Russian (FLEURS), Spanish (FLEURS)
- 1 hour of training data per language
- Pretrained model: MMS 300M (facebook/mms-300m)
- Full fine-tuning of all pretrained weights
- Additional attention head on top of the encoder embeddings for ASR
- CTC-only training (ctc_weight = 1, using ESPnet’s built-in CTC)
- Training for 40 epochs
- Batch size = 4 with gradient accumulation over 16 steps 

To control randomness, I set seed=0 in ESPnet. I also verified that batching, sorting, and the initial forward passes are deterministic: the per-sample CTC losses are identical across runs up until the first backward pass (i.e., batch 17 in epoch 1).

However, I observe significant variability in final performance across runs. In the case of Japanese, the difference between the best and worst CER over 5 runs is approximately 19 percentage points. I am attaching a plot showing the CERs for the five runs, along with tables summarizing the results.

I understand that CTC loss may exhibit non-deterministic behavior in the backward pass (depending on the implementation and hardware), but I would like to ask:

- Is this level of variability expected in this setup?
- Are there known sources of non-determinism in ESPnet (or PyTorch CTC) that could explain differences of this magnitude?
- Are there recommended practices to reduce this variability (e.g., specific flags, deterministic settings, or alternative implementations)?

Any guidance would be greatly appreciated

<img width="989" height="490" alt="Image" src="https://github.com/user-attachments/assets/c79632fa-9b5c-4aee-9103-280274cfd385" />
<img width="1278" height="730" alt="Image" src="https://github.com/user-attachments/assets/3af1a5f1-6d7b-41c4-937b-9515956e9139" />

**Comment by Fhrozen** [comment]
@dosubot

**Comment by sw005320** [comment]
@josecruzado21, can you share the results of the Japanese ASR? (`results.txt`) We may identify the issue.

@Shunsuke32, do you also observe large variances in a specific language (e.g., Japanese)?

**Comment by Shunsuke32** [comment]
I will definitely check if I observe similar variance for the same language. Our GPU cluster is currently down for repairs, so I'll investigate and report back once it's online!

As a side note, my setup is slightly different: I am comparing different architectures for downstream tasks (not running same-seed trials) and keeping the MMS 1B model frozen.

**Comment by josecruzado21** [comment]
@sw005320 

Thanks for replying! 
Sure, I am attaching the CER result.txt for dev and test for each of the 5 experiments I ran

[result_dev_exp1.txt](https://github.com/user-attachments/files/28315685/result_dev_exp1.txt)
[result_dev_exp2.txt](https://github.com/user-attachments/files/28315686/result_dev_exp2.txt)
[result_dev_exp3.txt](https://github.com/user-attachments/files/28315691/result_dev_exp3.txt)
[result_dev_exp4.txt](https://github.com/user-attachments/files/28315689/result_dev_exp4.txt)
[result_dev_exp5.txt](https://github.com/user-attachments/files/28315694/result_dev_exp5.txt)
[result_test_exp1.txt](https://github.com/user-attachments/files/28315688/result_test_exp1.txt)
[result_test_exp2.txt](https://github.com/user-attachments/files/28315693/result_test_exp2.txt)
[result_test_exp3.txt](https://github.com/user-attachments/files/28315687/result_test_exp3.txt)
[result_test_exp4.txt](https://github.com/user-attachments/files/28315692/result_test_exp4.txt)
[result_test_exp5.txt](https://github.com/user-attachments/files/28315690/result_test_exp5.txt)

**Comment by sw005320** [comment]
@josecruzado21, thanks for sharing the results.
I found that the large variations in Japanese stem from a mismatch in the script.
With this phenomenon, I kind of understand that the current might happen with tiny changes.
We should still identify the issues and fix them, but my guess is that this variation comes from the nature of the Japanese language, and reducing the variance with the consistent script would be more of a research problem.

**Comment by sw005320** [comment]
@josecruzado21, btw, did you also follow https://github.com/espnet/espnet/issues/6433#issuecomment-4410578269?
They are exactly what I was thinking (deterministic flags and identifying the atomic operations).

**Comment by josecruzado21** [comment]
Hi @sw005320, thank you very much for taking the time to look into it, I really appreciate it.

Can you elaborate more on the "mismatch in the script" reason? I am not very familiar with the Japanese language. My concern with the large variability is that I am comparing the performance of 2 update rules on the worst-performing language in the group (which in this case was Japanese). From what I understand from your comment, this behavior is very specific to Japanese because of its writing systems. I was concerned that this result might extrapolate to "worst-performing languages" more generally.

Regarding determinism in comment 6433, I tried some things:

- I realized CUDA implementation of CTC does not have a deterministic implementation, so `torch.use_deterministic_algorithms(True)` throws errors.
- The cuDNN implementation of CTC loss does have a deterministic implementation that is called if certain criteria are met (for instance, target_lengths <= 256, blank token is 0, and some others). I changed my script so these criteria are met, and my code started to effectively use cuDNN, but even then, results start to diverge across runs after the first backward pass.
- I would probably need to go much lower level to guarantee deterministic behavior. However, I am not so much bothered about irreproducibility as I am about the level of variability in the metrics.

Something the comment mentions (which I was also thinking about) is that 1 hour of data might be too small, making results overly sensitive to gradient noise for a full fine-tune of a 300M parameter model. Any thoughts on this?

In the meantime, I will try different language sets and increase the data size.

Thanks again for taking the time, I really appreciate your help

**Comment by sw005320** [comment]
> Hi [@sw005320](https://github.com/sw005320), thank you very much for taking the time to look into it, I really appreciate it.
> 
> Can you elaborate more on the "mismatch in the script" reason? I am not very familiar with the Japanese language. My concern with the large variability is that I am comparing the performance of 2 update rules on the worst-performing language in the group (which in this case was Japanese). From what I understand from your comment, this behavior is very specific to Japanese because of its writing systems. I was concerned that this result might extrapolate to "worst-performing languages" more generally.
> 

The ASR outputs often switch to other scripts (e.g., the Roman alphabet suddenly appears in the second position "コ **N** シ ズ") rather than Japanese. (Other examples in the other scripts at the 1st and 4th positions: "מ で あ ר る"). Even in Japanese, the use of Hiragana/Katakana/Kanji scripts is inconsistent, e.g., "場 に **フ サ** し い" shoud be "場 に ふ  さ. わ し い"). 
This script inconsistently increases the variations of the errors.
So, for me, your result is understandable. 
Again, we still need to work on fixing this, but fixing the script consistency in this range of the training data is another research topic for me.

**Comment by sw005320** [comment]
I'm revisiting this issue.

**Comment by sw005320** [comment]
@josecruzado21, I created the PR to make espnet as deterministic as possible in https://github.com/espnet/espnet/pull/6630
However, as you can see in the comments, it is impossible to remove the non-deterministic operations in the current major CUDA-based environments, so your problem may remain.
If you test it and find it useful (mitigate some randomness issues), that is great. I'll merge this PR to the main. 
But I'm expecting that this may not fully solve your issue.

**Comment by josecruzado21** [comment]
Hi @sw005320,

Thanks for looking into this.

I reran my earlier experiment in the exact same setting, now including the changes from #6630. I also added a new experiment using the full amount of ML-SUPERB data available for each language, instead of capping training data at 1hr per language as in my original setup.

Observations:

- Full-data setting: Japanese CER variability is noticeably reduced after the fix.

- 1hr-per-language setting: 4 out of 5 runs show reduced variability for Japanese after the fix. However, one run is a massive outlier (spiking to ~177% jpn CER, far above the rest of experiments). I don't think this outlier is attributable to the PR's changes. It looks more like an anomalous training divergence caused by CTC's non-determinism, which likely sent that particular Japanese run down a different (and much worse) training trajectory. I'm including the .txt files with the results; the problem appears to be that in this specific run, the model fails to recognize Japanese and instead outputs characters from other languages.

- Hebrew shows slightly worse variability after the fix, in both the 1hr and full-data settings.

- The remaining languages (deu, eng, rus, spa) are unchanged.

I'm only including Dev set results here, but Test set results follow a similar pattern.

The before/after comparison is shown below.

### Results after training in 1hr of data per language

<img width="472" height="245" alt="Image" src="https://github.com/user-attachments/assets/f1c09fc2-1f9d-4dd1-af0f-7208d354bb44" />

<img width="989" height="490" alt="Image" src="https://github.com/user-attachments/assets/b1f54602-f3a9-48fc-be54-e8dff3fcf390" />

### Results after training in all available data in ML SUPERB

<img width="472" height="248" alt="Image" src="https://github.com/user-attachments/assets/56ef7a11-b495-4652-9b76-4d0845476265" />

<img width="989" height="490" alt="Image" src="https://github.com/user-attachments/assets/19b38b5d-3b8f-44d9-ba9d-877ea66fa9f3" />

[result_exp4_jpn_outlier.txt](https://github.com/user-attachments/files/31976327/result_exp4_jpn_outlier.txt)

### Conclusions

The randomness control introduced in the PR reduces variability in Japanese, but the residual non-determinism from CTC remains problematic. This is especially evident in low-data regimes, where it can still push training toward suboptimal trajectories.
